APE26.rst

```diff
@@ -0,0 +1,325 @@
+cases (checking ``frame.has_data``), complicating the codebase.
+Moreover static analyzers cannot determine which case is being
+used; with separated frames and data static analyzers will be
+able to prevent this entire class of errors. We thus find it
+worthwhile to separate coordinate data from reference frames at
+the possible expense of developers having to learn this new
+framework.
+
+All the points discussed thus far – separation of concerns and
+code duplication – concern maintainers. However user experience
+is the more important consideration. In this arena too,
+separating frames from data storage has its advantages. Perhaps
+most importantly, documentation will be more obvious: the methods
+and attributes are defined on ``SkyCoord`` (and ``Coordinate``)
+proper, so sphinx will know how to typeset those, while type
+checkers can help users in finding and using them properly. It
+will also be easier: following the Zen of Python, there should be
+one clear way to do something. The present overlap leads to
+confusion wherein beginner users end up creating
+``BaseCoordinateFrame`` objects such as ``ICRS``, when the docs
+are clear that these are for more advanced users and that
+``SkyCoord`` is to be preferred. The system will also be less
+fragile. For example, if users manipulate the internal workings
+of ``SkyCoord`` (which is discouraged but possible), the
+coordinate data can become decoupled from the caching that
+``SkyCoord`` performs for speed. With these proposed changes
+users will have a more clear, introspectable, and robust system.
+
+Finished Product
+----------------
+The end result of the implementation of this APE are be two separate
+hierarchies of classes: reference frame classes and coordinate
+classes which bring together a reference frame and coordinate
+data. We discuss each class type in turn.
+
+Coordinate frame classes only hold information pertaining to
+the reference frame they represent and never actual coordinate
+data in that reference frame. This is consistent with our
+mathematical framework, the reference frame mediates how
+coordinate data is understood (e.g. distance measures) or
+interacts (e.g. separation from other coordinates), but the
+coordinate data itself is actually independent of that
+information.
+
+Classes like ``SkyCoord`` will be composed structures bringing
+together the reference frame (an instance of a ``BaseFrame``
+subclass) and the coordinate data (``BaseRepresentation``
+objects). We also introduce a new class, ``Coordinate``, which
+is akin to ``SkyCoord`` (containing both frame and data), but
+without extra features like keeping frame attributes not
+associated with the current frame, caching and flexible input
+parsing. In this way ``Coordinate`` operates very similarly to
+the current ``BaseCoordinateFrame`` objects when they have
+data, and is meant to be their direct replacement in the new
+framework as well as a more lightweight and performant
```

Comment:
How does the new `Coordinate` class avoid code duplication and user confusion that this APE tries to solve? Why can't we just use `SkyCoord`?

Reply:
> avoid code duplication and user confusion that this APE tries to solve?

Common base class.

Reply:
Am I supposed to learn that from reading this paragraph?

Reply:
Yes, in part. See the lower pseudo-code that demonstrates how simple `Coordinate` will be as a subclass of the common base class. That and the assumption that for changes large enough to merit an APE there would need to be a reasonable amount of explanation. For example, even a `See Also` is very useful for helping users decide when to use a `Coordinate` versus a `SkyCoord`. Combined with further docstring and RTD changes which are the natural course of a PR, and I'm not so worried.